### PR TITLE
feat(sapling): support proof generation from provingkey

### PR DESCRIPTION
**Pull request by roxaneletourneau**
Implemented an InMemorySpendingKey class that allows producing proofa independently from the
spending key. Did some refactor and renaming also.

BREAKING CHANGE: The first parameter of the SaplingToolkit class changed for an object accepting an
optional InMemorySpendingKey

Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [ ] Your code builds cleanly without any errors or warnings
- [ ] You have added unit tests
- [ ] You have added integration tests (if relevant/appropriate)
- [ ] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

File: packages/taquito-sapling/src/sapling-keys/in-memory-proving-key.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  ): Promise<Omit<SaplingSpendDescription, 'signature'>> {
+    const spendDescription = await sapling.prepareSpendDescriptionWithAuthorizingKey(
+      parametersSpendProof.saplingContext,
+      this.#provingKey,
+      parametersSpendProof.address,
+      parametersSpendProof.randomCommitmentTrapdoor,
+      parametersSpendProof.publicKeyReRandomization,
+      parametersSpendProof.amount,
+      parametersSpendProof.root,
+      parametersSpendProof.witness
```
Comment (zainen) [inline packages/taquito-sapling/src/sapling-keys/in-memory-proving-key.ts:57]: curious what the reason is to deconstruct  `parametersSpendProof`  to pass each individually here

File: packages/taquito-sapling/src/sapling-tx-builder/sapling-transactions-builder.ts [hunk 2]
```diff
@@ -23,9 +23,11 @@ import { convertValueToBigNumber } from '../sapling-tx-viewer/helpers';
 import { SaplingState } from '../sapling-state/sapling-state';
 import { SaplingWrapper } from '../sapling-module-wrapper';
 import { InMemorySpendingKey } from '../sapling-keys/in-memory-spending-key';
+import { InMemoryProvingKey } from '../sapling-keys/in-memory-proving-key';
 
 export class SaplingTransactionBuilder {
   #inMemorySpendingKey: InMemorySpendingKey;
+  #inMemoryProvingKey: InMemoryProvingKey | undefined;
```
Comment (zainen) [inline packages/taquito-sapling/src/sapling-tx-builder/sapling-transactions-builder.ts:30]: does this also work as `#inMemoryProvingKey?: InMemoryProvingKey` ?

File: packages/taquito-sapling/src/taquito-sapling.ts [hunk 3]
```diff
@@ -30,15 +30,18 @@ import {
 } from './types';
 import { SaplingTransactionBuilder } from './sapling-tx-builder/sapling-transactions-builder';
 import { DEFAULT_BOUND_DATA, DEFAULT_MEMO } from './constants';
+import { InMemoryProvingKey } from './sapling-keys/in-memory-proving-key';
 
 export { SaplingTransactionViewer } from './sapling-tx-viewer/sapling-transaction-viewer';
 export { InMemoryViewingKey } from './sapling-keys/in-memory-viewing-key';
 export { InMemorySpendingKey } from './sapling-keys/in-memory-spending-key';
+export { InMemoryProvingKey } from './sapling-keys/in-memory-proving-key';
```
Comment (zainen) [inline packages/taquito-sapling/src/taquito-sapling.ts:38]: are these two imports of the same thing?

File: packages/taquito-sapling/src/taquito-sapling.ts [hunk 4]
```diff
@@ -30,15 +30,18 @@ import {
...
 
 /**
  * @description Class that surfaces all of the sapling capability allowing to read from a sapling state and prepare transactions
  *
- * @param inMemorySpendingKey Holds the sapling spending key
+ * @param keys.saplingSigner Holds the sapling spending key
+ * @param keys.saplingProver Optional Allows to generate the proofs with the proving key rather than the spending key
```
Comment (zainen) [inline packages/taquito-sapling/src/taquito-sapling.ts:44]: suggestion `(Optional) Allows the generation of proofs`

**Review by zainen — COMMENTED** [review]
LGTM nice refactors of InMemorySpendingKey! just a couple questions or suggestions

File: packages/taquito-sapling/src/taquito-sapling.ts (hunk 3 above)
Comment (roxaneletourneau) [reply packages/taquito-sapling/src/taquito-sapling.ts:38]: The first one is an import and the second an export. In the main file of the package, we need to explicitly export all classes that we want to make available to the users.

File: packages/taquito-sapling/src/sapling-keys/in-memory-proving-key.ts (hunk 1 above)
Comment (roxaneletourneau) [reply packages/taquito-sapling/src/sapling-keys/in-memory-proving-key.ts:57]: sapling.prepareSpendDescriptionWithAuthorizingKey from '@airgap/sapling-wasm doesn't take an object as parameter but individual arguments

File: packages/taquito-sapling/src/sapling-tx-builder/sapling-transactions-builder.ts (hunk 2 above)
Comment (roxaneletourneau) [reply packages/taquito-sapling/src/sapling-tx-builder/sapling-transactions-builder.ts:30]: yes

File: integration-tests/sapling-transactions-contract-with-single-state.spec.ts [hunk 5]
```diff
@@ -32,21 +32,21 @@ CONFIGS().forEach(({ lib, rpc, setup }) => {
       const mnemonic: string = bip39.generateMnemonic();
       bobInmemorySpendingKey = await InMemorySpendingKey.fromMnemonic(mnemonic);
 
-      // Instantiate an InMemorySpendingKey from a spending key for Alice
-      aliceInmemorySpendingKey = new InMemorySpendingKey('sask27SLmU9herddHz4qFJBLMjWYMbJF8RtS579w9ej9mfCYK7VUdyCJPHK8AzW9zMsopGZEkYeNjAY7Zz1bkM7CGu8eKLzrjBLTMC5wWJDhxiK91ahA29rhDRsHdJDV2u2jFwb2MNUix8JW7sAkAqYVaJpCehTBPgRQ1KqKwqqUaNmuD8kazd4Q8MCWmgbWs21Yuomdqyi9FLigjRp7oY4m5adaVU19Nj1AHvsMY2tePeU2L')
+      // Instanciate an InMemorySpendingKey from a spending key for Alice
```
Comment (dsawali) [inline integration-tests/sapling-transactions-contract-with-single-state.spec.ts:35]: I think instantiate was correct 😄

**Review by dsawali — CHANGES_REQUESTED** [review]

File: packages/taquito-sapling/src/sapling-forger/sapling-forger.ts [hunk 6]
```diff
@@ -57,7 +57,7 @@ export class SaplingForger {
     return Buffer.concat([
       desc.commitmentValue,
       desc.nullifier,
-      desc.randomizedPublicKey,
+      desc.publicKeyReRandomization,
```
[the remark this answers is not in the document]
Comment (roxaneletourneau) [reply packages/taquito-sapling/src/sapling-forger/sapling-forger.ts:60]: because I used the different names at different places

**Pull request merged**